// kitsu-core/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var deattribute = require('./deattribute.js');
var error = require('./error.js');
var filterIncludes = require('./filterIncludes.js');
var linkRelationships = require('./linkRelationships.js');
var deserialise = require('./deserialise.js');
var query = require('./query.js');
var serialise = require('./serialise.js');
var camel = require('./camel.js');
var kebab = require('./kebab.js');
var snake = require('./snake.js');

function splitModel(url) {
  let options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.pluralModel) options.pluralModel = s => s;
  if (!options.resourceCase) options.resourceCase = s => s;
  const urlSegments = url.split('/');
  const resourceModel = urlSegments.pop();
  urlSegments.push(options.pluralModel(options.resourceCase(resourceModel)));
  const newUrl = urlSegments.join('/');
  return [resourceModel, newUrl];
}

exports.deattribute = deattribute.deattribute;
exports.error = error.error;
exports.filterIncludes = filterIncludes.filterIncludes;
exports.linkRelationships = linkRelationships.linkRelationships;
exports.deserialise = deserialise.deserialise;
exports.query = query.query;
exports.serialise = serialise.serialise;
exports.camel = camel;
exports.kebab = kebab;
exports.snake = snake;
exports.splitModel = splitModel;
